compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/InterfaceLowering.kt

```diff
@@ -22,59 +25,95 @@ import org.jetbrains.kotlin.ir.expressions.IrReturn
 
         val defaultImplsIrClass = context.declarationFactory.getDefaultImplsClass(irClass)
-        //TODO: Don't add DefaultImpls if it's empty. Just like the old backend...?
-        irClass.declarations.add(defaultImplsIrClass)
         val members = defaultImplsIrClass.declarations
 
-        for (function in irClass.declarations) {
-            if (function !is IrSimpleFunction) continue
-
-            if (function.modality != Modality.ABSTRACT && function.origin != IrDeclarationOrigin.FAKE_OVERRIDE) {
-
-                if(function.hasJvmDefault() && !context.state.jvmDefaultMode.isCompatibility && !mustMoveToDefaultImpls(function)) continue
+        // There are 6 cases for functions on interfaces:
+        loop@ for (function in irClass.functions) {
+            when {
+                /**
+                 * 1) They are plain abstract interface functions, in which case we leave them:
+                 */
+                function.modality == Modality.ABSTRACT ->
+                    continue@loop
+
+                /**
+                 * 2) They inherit a default implementation from an interface this interface
+                 *    extends: create a bridge from companion to companion, if necessary:
+                 *
+                 *    ```
+                 *    interface A { fun foo() = 0 }
+                 *    interface B : A { }
+                 *    ```
+                 *
+                 *    yields
+                 *
+                 *    ```
+                 *    interface A { fun foo(); class DefaultImpls { fun foo() = 0 } }  // !! by Case 4 !!
+                 *    interface B : A { class DefaultImpls { fun foo() = A.DefaultImpls.foo() } }
+                 *    ```
+                 */
+                function.origin == IrDeclarationOrigin.FAKE_OVERRIDE -> {
+                    val implementation = function.resolveFakeOverride()!!
+
+                    if (!Visibilities.isPrivate(implementation.visibility)
+                        && !implementation.isMethodOfAny()
+                        && (!implementation.hasJvmDefault() || context.state.jvmDefaultMode.isCompatibility)
+                    ) {
+                        delegateInheritedDefaultImplementationToDefaultImpls(function, implementation, defaultImplsIrClass)
+                    }
+                }
 
-                val element = context.declarationFactory.getDefaultImplsFunction(function).also {
-                    if (mustMoveToDefaultImpls(function))
-                        removedFunctions[function.symbol] = it.symbol
+                /**
+                 * 3) Private methods, default parameter dispatchers (without @JvmDefault)
+                 *    and $annotation methods are always moved without bridges
+                 */
+                Visibilities.isPrivate(function.visibility)
+                        || (function.origin == IrDeclarationOrigin.FUNCTION_FOR_DEFAULT_PARAMETER && !function.hasJvmDefault())
```

Comment:
@max-kammerer This is the point you made earlier, no?

Reply:
Look like yes (cause of changed code). But it's not clear is JvmDefault annotation presented on $default companion (shouldn't we check original function)?

Reply:
Yep, [we copy all annotations on to the $default method](https://github.com/JetBrains/kotlin/blob/ea56a5e8b1eebcbd6648db2105712daf9ee7ace0/compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/DefaultArgumentStubGenerator.kt#L458).